cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i

```diff
@@ -0,0 +1,204 @@
+#include "src/algorithms/cosdistance/cosdistance_kernel.h"
+#include "src/externals/service_lapack.h"
+#include <iostream>
+
+
+using namespace daal::internal;
+
+namespace daal
+{
+namespace algorithms
+{
+namespace spectral_embedding
+{
+namespace internal
+{
+
+template <typename algorithmFPType, CpuType cpu>
+services::Status computeEigenvectorsInplace(size_t nFeatures, algorithmFPType * eigenvectors, algorithmFPType * eigenvalues)
+{
+    char jobz = 'V';
+    char uplo = 'U';
+
+    DAAL_INT lwork  = 2 * nFeatures * nFeatures + 6 * nFeatures + 1;
+    DAAL_INT liwork = 5 * nFeatures + 3;
+    DAAL_INT info;
+
+    TArray<algorithmFPType, cpu> work(lwork);
+    TArray<DAAL_INT, cpu> iwork(liwork);
+    DAAL_CHECK_MALLOC(work.get() && iwork.get());
+
+    LapackInst<algorithmFPType, cpu>::xsyevd(&jobz, &uplo, (DAAL_INT *)(&nFeatures), eigenvectors, (DAAL_INT *)(&nFeatures), eigenvalues, work.get(),
+                                             &lwork, iwork.get(), &liwork, &info);
+    if (info != 0) return services::Status(services::ErrorPCAFailedToComputeCorrelationEigenvalues); // CHANGE ERROR STATUS
+    return services::Status();
+}
+
+
+/**
+ *  \brief Kernel for Spectral Embedding calculation
+ */
+template <typename algorithmFPType, Method method, CpuType cpu>
+services::Status SpectralEmbeddingKernel<algorithmFPType, method, cpu>::compute(const NumericTable* xTable, NumericTable* eTable, const KernelParameter & par)
+{
+    
+    services::Status safeStat;
+    std::cout << "inside DAAL kernel" << std::endl;
+    std::cout << "Params: " << par.numEmb << " " << par.numNeighbors << std::endl;
+    size_t k = par.numEmb;
+    size_t filtNum = par.numNeighbors + 1;
+    size_t n = xTable->getNumberOfRows();    /* Number of input feature vectors   */
+
+    SharedPtr<HomogenNumericTable<algorithmFPType> > tmpMatrixPtr = HomogenNumericTable<algorithmFPType>::create(n, n, NumericTable::doAllocate, &safeStat);
+    if (!safeStat) {
+        return safeStat;
+    }
+    NumericTable* covOutput = tmpMatrixPtr.get();
+
+    NumericTable * a0                      = const_cast<NumericTable *>(xTable);
+    NumericTable ** a                      = &a0;
+    NumericTable ** r                      = &covOutput;
+    
+
+    // Compute cosine distances matrix
+    auto* cosDistKernelPtr = new cosine_distance::internal::DistanceKernel<algorithmFPType, cosine_distance::Method::defaultDense, cpu>();
+    
+    services::Status cosDistStatus = cosDistKernelPtr->compute(0, a, 0, r, nullptr);
+
+    delete cosDistKernelPtr;
+
+    if (!cosDistStatus) {
+        return cosDistStatus;
+    }
+
+    WriteRows<algorithmFPType, cpu> xMatrix(*r, 0, n);
+    DAAL_CHECK_BLOCK_STATUS(xMatrix);
+    algorithmFPType * x = xMatrix.get();
+
+    size_t lcnt, rcnt, cnt;
+    algorithmFPType L, R, M;
+    // Use binary search to find such d that the number of verticies having distance <= d is filtNum
+    const size_t binarySearchIterNum = 20;
```

Comment:
What means a magic number 20?

Reply:
For now it is actually a magic number but it should be log(1/eps) where eps is floating point number precision

Reply:
It would be better then to define pre-computed constants for `float` and `double`, add a description, and use those constants in the computations.
For example, like it is done here: https://github.com/oneapi-src/oneDAL/blob/main/cpp/daal/src/algorithms/em/em_gmm_dense_default_batch_impl.i#L144
lines 144-157